File: Solutions/TwoSum.cpp
// Leetcode Problem #1: https://leetcode.com/problems/two-sum/
// Source: https://leetcode.com/problems/two-sum/discuss/13/Accepted-C++-O(n)-Solution/263

// Solution 1: Map | O(n) time, O(n) space
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // Create a map of values w/ indices of the elements "looking for it"
        unordered_map<int, int> map;
        
        // For each number in nums
        for (int i = 0; i < nums.size(); i++) {
            // If the element is in the map, return the indices
            auto it = map.find(nums[i]);
            if (it != map.end()) return {it->second, i};

            // Else store target - current element with the current index
            map[target - nums[i]] = i;
        }
        
        return {}; // Solution not found, return empty vector. SHOULD NOT HAPPEN!
    }
};

File: Solutions/TwoSum.py
# Leetcode #2: https://leetcode.com/problems/two-sum/description/
# Source: https://leetcode.com/problems/two-sum/solutions/127810/two-sum/

# Solution 1: Hash Map | O(n) time, O(n) space
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Create a dictionary to map elements with their indices
        hashMap = {}

        # Loop through the list and check if the complement is in the hash map
        # If yes, return the 2 indices. If no, add current number to the hash map
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in hashMap:
                return [hashMap[complement], i]
            hashMap[nums[i]] = i
